sdks/go/pkg/beam/core/runtime/xlangx/expand.go

```diff
@@ -140,3 +142,67 @@ func QueryExpansionService(ctx context.Context, p *HandlerParams) (*jobpb.Expans
 	}
 	return res, nil
 }
+
+func startAutomatedExpansionService(gradleTarget string) (func() error, string, error) {
+	jarPath, err := expansionx.GetBeamJar(gradleTarget, core.SdkVersion)
+	if err != nil {
+		return nil, "", err
+	}
+	serviceRunner, err := expansionx.NewExpansionServiceRunner(jarPath, "")
+	if err != nil {
+		return nil, "", err
+	}
+	err = serviceRunner.StartService()
+	if err != nil {
+		return nil, "", err
+	}
+	return serviceRunner.StopService, serviceRunner.Endpoint(), nil
+}
+
+// QueryAutomatedExpansionService submits an external transform to be expanded by the
+// expansion service and then eagerly materializes the artifacts for staging. The given
+// transform should be the external transform, and the components are any additional
+// components necessary for the pipeline snippet.
+//
+// The address to be queried is determined by the Config field of the HandlerParams after
+// the prefix tag indicating the automated service is in use.
+func QueryAutomatedExpansionService(ctx context.Context, p *HandlerParams) (*jobpb.ExpansionResponse, error) {
+	// Strip auto: tag to get Gradle target
+	tag, target := parseAddr(p.Config)
+
+	stopFunc, address, err := startAutomatedExpansionService(target)
+	if err != nil {
+		return nil, err
+	}
+	defer stopFunc()
+
+	p.Config = address
+
+	res, err := QueryExpansionService(ctx, p)
+	if err != nil {
+		return nil, err
+	}
+
+	exp := &graph.ExpandedTransform{
+		Components:   res.GetComponents(),
+		Transform:    res.GetTransform(),
+		Requirements: res.GetRequirements(),
+	}
+
+	p.ext.Expanded = exp
+	// Put correct expansion address into edge
+	p.edge.External.ExpansionAddr = address
```

Comment:
(No change required) Changing the address here and then changing it right back after feels a little hacky to me, but I don't see any good alternatives at the moment. I just think it's another sign that at some point we'll need to come back to this and polish it further (along with retaining expansion services so using the same transform multiple times doesn't result in handling duplicate jars and artifacts).

Reply:
Yeah it's a little hacky, but how we can avoid the second materialize() call easily.

Reply:
A subsequent PR can probably just add a flag to the graph.External node in graph, that we can check to prevent the later materialization.